### PR TITLE
[weight daemon] support draft model

### PR DESCRIPTION
**Pull request by liusy58**
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #32818301875](https://github.com/sgl-project/sglang/actions/runs/32818301875)
Latest PR Test (Extra): :white_check_mark: [Run #32818301638](https://github.com/sgl-project/sglang/actions/runs/32818301638)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32818301791](https://github.com/sgl-project/sglang/actions/runs/32818301791)

### CODE REVIEW

File: python/sglang/srt/entrypoints/engine.py [hunk 1]
```diff
@@ -701,50 +701,83 @@ def _launch_weight_cache_daemons(cls, server_args: ServerArgs):
             # engine's own NCCL TCPStore.
             dist_init_method = NetworkAddress("127.0.0.1", get_free_port()).to_tcp()
 
-        num_daemons = len(pp_rank_range) * len(tp_rank_range)
+        from sglang.srt.weight_cache.daemon import spawn_weight_cache_daemon
```
Comment (alexnails) [inline python/sglang/srt/entrypoints/engine.py]: I believe we had this for other PR, but let's not bring this import into hot path when not used?

File: python/sglang/srt/entrypoints/engine.py [hunk 2]
```diff
@@ -701,50 +701,83 @@ def _launch_weight_cache_daemons(cls, server_args: ServerArgs):
...
+
+        # Speculative decoding: the draft/MTP model is a separate nn.Module, so
+        # each rank gets a second daemon on its own "_draft" socket. The draft
+        # daemons form their own process group and need their own rendezvous
+        # port (single-node only; check_server_args rejects the multi-node
+        # combination).
+        draft_dist_init_method = None
+        if server_args.speculative_algorithm is not None:
+            draft_dist_init_method = NetworkAddress(
+                "127.0.0.1", get_free_port()
```
Comment (alexnails) [inline python/sglang/srt/entrypoints/engine.py]: should this be hardcoded like this? do we not have this abstracted somewhere?

File: python/sglang/srt/entrypoints/engine.py [hunk 3]
```diff
@@ -701,50 +701,83 @@ def _launch_weight_cache_daemons(cls, server_args: ServerArgs):
...
+            ).to_tcp()
```
Comment (alexnails) [inline python/sglang/srt/entrypoints/engine.py]: does this have to be to tcp?

File: python/sglang/srt/weight_cache/protocol.py [hunk 4]
```diff
@@ -274,6 +284,150 @@ def compute_env_stamp() -> Dict[str, str]:
...
+    """Build the per-rank daemon set: the target model, plus the draft model
+    when speculative decoding is enabled.
+
+    Model identity per role comes from ``resolve_daemon_model_identity``, the
+    same helper the daemon uses, so the launcher's view and the daemon's load
+    always describe the same checkpoint.
+    """
+    specs = [
+        DaemonModelSpec(
+            is_draft_model=False,
```
Comment (alexnails) [inline python/sglang/srt/weight_cache/protocol.py:410]: I do not think we can always assume this as default? I may be wrong

**Review by alexnails — COMMENTED** [review]
first pass

File: python/sglang/srt/entrypoints/engine.py (hunk 1 above)
Comment (liusy58) [reply python/sglang/srt/entrypoints/engine.py]: Kept the `heavy weight_cache.daemon` import right before the spawn loop;

File: python/sglang/srt/entrypoints/engine.py (hunk 2 above)
Comment (liusy58) [reply python/sglang/srt/entrypoints/engine.py]: Good catch, it was duplicated in three places. Added `make_local_rendezvous()` in weight_cache/protocol.py

File: python/sglang/srt/entrypoints/engine.py (hunk 3 above)
Comment (liusy58) [reply python/sglang/srt/entrypoints/engine.py]: Yes — the daemons form a torch process group via `init_distributed_environment`, which rendezvouses through TCPStore.

File: python/sglang/srt/weight_cache/protocol.py (hunk 4 above)
Comment (liusy58) [reply python/sglang/srt/weight_cache/protocol.py:410]: These are per-entry labels, not defaults: `specs[0]` is the target, `specs[1]` (appended only under speculative decoding) is the draft with True.
